src/grader/test_lib.mli

```diff
@@ -15,343 +15,843 @@
 
-  (** The type of arguments, represented as heterogeneous lists.
-
-  Usage: [arg 3 @@ arg "word" @@ last false] *)
-  type ('arrow, 'uarrow, 'ret) args
-  val last :
-    'a ->
-    ('a -> 'ret, 'a -> unit, 'ret) args
-  val arg :
-    'a ->
-    ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
-    ('a -> 'ar -> 'row, 'a -> 'ar -> 'urow, 'ret) args
-
-  val apply : ('ar -> 'row) -> ('ar -> 'row, 'ar -> 'urow, 'ret) args -> 'ret
-
-  (** The type of function prototypes.
-
-  Usage: [arg_ty [%ty: int] @@ arg_ty [%ty: string] @@ last_ty [%ty: bool] [%ty: unit]] *)
-  type ('arrow, 'uarrow, 'ret) prot
-  val last_ty :
-    'a Ty.ty ->
-    'ret Ty.ty ->
-    (('a -> 'ret) Ty.ty, 'a -> unit, 'ret) prot
-  val arg_ty :
-    'a Ty.ty ->
-    (('ar -> 'row) Ty.ty, 'ar -> 'urow, 'ret) prot ->
-    (('a -> 'ar -> 'row) Ty.ty, ('a -> 'ar -> 'urow), 'ret) prot
-
-  type 'a lookup = unit -> [ `Found of string * Learnocaml_report.report * 'a | `Unbound of string * Learnocaml_report.report ]
-
-  val lookup : 'a Ty.ty -> ?display_name: string -> string -> 'a lookup
-  val lookup_student : 'a Ty.ty -> string -> 'a lookup
-  val lookup_solution : 'a Ty.ty -> string -> 'a lookup
-  val found : string -> 'a -> 'a lookup
-  val name : 'a lookup -> string
-
-  val test_value : 'a lookup -> ('a -> Learnocaml_report.report) -> Learnocaml_report.report
-
-  val test_function :
-    ?test: 'ret tester ->
-    ?test_stdout: io_tester ->
-    ?test_stderr: io_tester ->
-    ?before :
-      (('ar -> 'row, 'ar -> 'urow, 'ret) args ->
-       unit) ->
-    ?after :
-      (('ar -> 'row, 'ar -> 'urow, 'ret) args ->
-       ('ret * string * string) ->
-       ('ret * string * string) ->
-       Learnocaml_report.report) ->
-    (('ar -> 'row) Ty.ty, 'ar -> 'urow, 'ret) prot ->
-    ('ar -> 'row) lookup ->
-    (('ar -> 'row, 'ar -> 'urow, 'ret) args * (unit -> 'ret)) list ->
-    Learnocaml_report.report
-
-  val test_function_against :
-    ?gen: int ->
-    ?test: 'ret tester ->
-    ?test_stdout: io_tester ->
-    ?test_stderr: io_tester ->
-    ?before_reference :
-      (('ar -> 'row, 'ar -> 'urow, 'ret) args -> unit) ->
-    ?before_user :
-      (('ar -> 'row, 'ar -> 'urow, 'ret) args -> unit) ->
-    ?after :
-      (('ar -> 'row, 'ar -> 'urow, 'ret) args ->
-       ('ret * string * string) ->
-       ('ret * string * string) ->
-       Learnocaml_report.report) ->
-    ?sampler:
-      (unit -> ('ar -> 'row, 'ar -> 'urow, 'ret) args) ->
-    (('ar -> 'row) Ty.ty, 'ar -> 'urow, 'ret) prot ->
-    ('ar -> 'row) lookup -> ('ar -> 'row) lookup ->
-    ('ar -> 'row, 'ar -> 'urow, 'ret) args list ->
-    Learnocaml_report.report
+    (** The various grading functions use numerous commun optional
+       argument. Here is a list in alphabetic order of each of them.
 
-  (*----------------------------------------------------------------------------*)
+    {3 ?⁠after} Defines a function which is called with the current
```

Comment:
If `Defines` has `?after` as a subject, no uppercase is needed. Otherwise, no 's' is needed at the end of `Defines`.